### PR TITLE
Add PyLong Import/Export API

### PR DESCRIPTION
**Pull request by skirpichev**
PyPy is not supported, as well as Py2.  In the later case it's possible, but hardly worth code complications: most real-word potential consumers (e.g. Sage, gmpy2 or python-flint) support only Python 3.

### CODE REVIEW

**Comment by vstinner** [comment]
https://github.com/python/cpython/pull/121339 was merged. What's the status of this PR? Is it up to date and ready to be reviewed?

File: pythoncapi_compat.h [hunk 1]
```diff
@@ -1720,6 +1720,187 @@ static inline int PyLong_AsUInt64(PyObject *obj, uint64_t *pvalue)
...
+// Helpers to access PyLongObject internals.
+static inline void
+_PyLong_SetSignAndDigitCount(PyLongObject *op, int sign, Py_ssize_t size)
+{
+#if PY_VERSION_HEX >= 0x030C0000
+    op->long_value.lv_tag = (uintptr_t)(1 - sign) | ((uintptr_t)(size) << 3);
+#elif PY_VERSION_HEX >= 0x030900A4
+    Py_SET_SIZE(op, sign*size);
+#else
+    Py_SIZE(op) = sign*size;
```
Comment (vstinner) [inline pythoncapi_compat.h:1734]:
```suggestion
    Py_SIZE(op) = sign * size;
```

File: pythoncapi_compat.h [hunk 2]
```diff
@@ -1720,6 +1720,187 @@ static inline int PyLong_AsUInt64(PyObject *obj, uint64_t *pvalue)
...
+}
+
+typedef struct PyLongLayout {
+    uint8_t bits_per_digit;
+    uint8_t digit_size;
+    int8_t digits_order;
+    int8_t digit_endianness;
+} PyLongLayout;
+
+static const PyLongLayout PyLong_LAYOUT = {
```
Comment (vstinner) [inline pythoncapi_compat.h:1765]: Is it possible to move it inside PyLong_GetNativeLayout()?

File: pythoncapi_compat.h [hunk 3]
```diff
@@ -1720,6 +1720,187 @@ static inline int PyLong_AsUInt64(PyObject *obj, uint64_t *pvalue)
...
+        --i;
+    }
+    if (i != j) {
+        if (i == 0) {
+            sign = 0;
+        }
+        _PyLong_SetSignAndDigitCount(self, sign, i);
+    }
+    if (i <= 1) {
+        long val = sign*(long)(_PyLong_GetDigits(self)[0]);
```
Comment (vstinner) [inline pythoncapi_compat.h:1894]:
```suggestion
        long val = sign * (long)(_PyLong_GetDigits(self)[0]);
```

File: tests/test_pythoncapi_compat_cext.c [hunk 4]
```diff
@@ -1426,6 +1426,50 @@ test_long_api(PyObject *Py_UNUSED(module), PyObject *Py_UNUSED(args))
...
+    digit *digits;
+    PyLongWriter *writer;
+    static PyLongExport long_export;
+
+    writer = PyLongWriter_Create(1, 1, (void**)&digits);
+    PyLongWriter_Discard(writer);
+
+    writer = PyLongWriter_Create(1, 1, (void**)&digits);
+    digits[0] = 123;
+    obj = PyLongWriter_Finish(writer);
```
Comment (vstinner) [inline tests/test_pythoncapi_compat_cext.c:1440]: I suggest adding: `check_int(obj, -123);`.

File: tests/test_pythoncapi_compat_cext.c [hunk 5]
```diff
@@ -1426,6 +1426,50 @@ test_long_api(PyObject *Py_UNUSED(module), PyObject *Py_UNUSED(args))
...
+    assert(digits[0] == 1);
+    assert(digits[1] == 0);
+    assert(digits[2] == 0);
+    assert(digits[3] == 0);
+    assert(digits[4] == 1);
+    PyLong_FreeExport(&long_export);
+    Py_DECREF(obj);
+
+    const PyLongLayout *layout = PyLong_GetNativeLayout();
+
```
Comment (vstinner) [inline tests/test_pythoncapi_compat_cext.c:1468]: You may remove this empty line.

File: docs/changelog.rst [hunk 6]
```diff
@@ -1,6 +1,18 @@
 Changelog
 =========
 
+* XXXX-XX-XX: Add functions and structs:
```
Comment (skirpichev) [inline docs/changelog.rst:4]:
```suggestion
* 2024-12-13: Add functions and structs:
```

**Review by skirpichev — COMMENTED** [review]
> Is it up to date and ready to be reviewed?

It's now slightly outdated.  PyLongWriter_Create() should reject ndigits==0; export_long should be set to 0 on errors.  I'll address this today.

(Unfortunately, I still don't know how implement this for PyPy (not in RPython, but within this project).)

File: tests/test_pythoncapi_compat_cext.c (hunk 5 above)
Comment (skirpichev) [inline tests/test_pythoncapi_compat_cext.c:1468]:
```suggestion
```

**Comment by vstinner** [comment]
Merged, thank you.

**Pull request merged**